consul/consul.go

```diff
@@ -1,97 +1,52 @@
 package consul
 
 import (
+	"github.com/alaa/pencil-go/registry"
 	consul "github.com/hashicorp/consul/api"
 )
 
-type ConsulClient struct {
-	client *consul.Client
+// ServiceRepository is consul-based implementation of registry.ServiceRepository
+type ServiceRepository struct {
+	consulAgent consulAgent
 }
 
-func NewConsulClient() (ConsulClient, error) {
-	client, err := consul.NewClient(consul.DefaultConfig())
-	if err != nil {
-		return ConsulClient{}, err
-	}
-	return ConsulClient{client: client}, nil
-}
-
-type ConsulAgent struct {
-	agent *consul.Agent
+type consulAgent interface {
+	Services() (map[string]*consul.AgentService, error)
+	ServiceRegister(service *consul.AgentServiceRegistration) error
+	ServiceDeregister(serviceID string) error
 }
 
-func (c *ConsulClient) NewConsulAgent() ConsulAgent {
-	agent := c.client.Agent()
-	return ConsulAgent{agent: agent}
-}
-
-type Member struct {
-	Name string
-	IP   string
-	Port uint16
-}
-
-type Members []Member
-
-func buildMember(name string, ip string, port uint16) Member {
-	return Member{Name: name, IP: ip, Port: port}
-}
-
-func (a *ConsulAgent) members() Members {
-	list := Members{}
-	use_wan := false
-	members, _ := a.agent.Members(use_wan)
-	for _, member := range members {
-		list = append(list, buildMember(member.Name, member.Addr, member.Port))
-	}
-	return list
+// NewServiceRepository creates new instance of ServiceRepository structure
+func NewServiceRepository(consulAgent consulAgent) *ServiceRepository {
+	return &ServiceRepository{consulAgent}
 }
 
-func buildService(id string, name string, port int, ip string) consul.AgentServiceRegistration {
-	return consul.AgentServiceRegistration{ID: id, Name: name, Port: port, Address: ip}
+// Register adds service into consul
+func (c *ServiceRepository) Register(service *registry.Service) error {
+	return c.consulAgent.ServiceRegister(
+		c.serviceToAgentServiceRegistration(service),
+	)
 }
 
-func (a *ConsulAgent) Services() (map[string]*consul.AgentService, error) {
-	services, err := a.agent.Services()
-	if err != nil {
-		return services, err
-	}
-	return services, nil
-}
-
-func (a *ConsulAgent) ServicesIDs() ([]string, error) {
-	services, err := a.agent.Services()
-	if err != nil {
-		return []string{""}, err
-	}
-
-	list := []string{}
-	for _, srv := range services {
-		list = append(list, srv.ID)
-	}
-
-	return list, nil
-}
-
-func (a *ConsulAgent) RegisterService(id string, name string, port int, ip string) error {
-	srv := buildService(id, name, port, ip)
-	if err := a.agent.ServiceRegister(&srv); err != nil {
-		return err
-	}
-	return nil
+// Unregister removes service from consul
+func (c *ServiceRepository) Unregister(serviceID string) error {
```

Comment:
I used name `Unregister` to conform to `registry.ServiceRepository` interface. I don't see a problem with changing it to `Deregister` but do you think that because some implementation detail (in this case consul) has `Deregister` method our higher level interface should has method with the same name?
What if we'll replace consul with something other that has for example `Delete` method instead of `Deregister`? :)